cloudquotas: resolved permadiff for `preferred_value` in `google_cloud_quotas_quota_preference`

ramonvermeulen opened the pull request:
Closes https://github.com/hashicorp/terraform-provider-google/issues/24480
<!--
Complete the self-review checklist to help speed up the review process: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

If your PR is still work in progress, please create it in draft mode.

Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to.
For example: Fixes https://github.com/hashicorp/terraform-provider-google/issues/ISSUE_ID
-->

**Release Note Template for Downstream PRs (will be copied)**

See [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) for guidance.

```release-note:bug
cloudquotas: resolved permadiff for `preferred_value` in `google_cloud_quotas_quota_preference`
```

NickElliot reviewed (changes requested):
this would not be the ideal approach here:
would be better off making a diff suppress func that will suppress the diff on update if the plan is null->0

ramonvermeulen commented:
> this would not be the ideal approach here: would be better off making a diff suppress func that will suppress the diff on update if the plan is null->0

Check, I will change this to a diff_suppress_func. I went with custom_flatten because of the documentation on diff_suppress_func states:

_"Specifies the name of a [diff suppress function](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-behaviors#diffsuppressfunc) to use for this field. **In many cases, a [custom flattener](https://googlecloudplatform.github.io/magic-modules/develop/custom-code/#custom_flatten) is preferred because it will allow the user to see a clearer diff** when the field actually is being changed."_

ramonvermeulen commented:
Just changed to a diff_suppress_func in https://github.com/GoogleCloudPlatform/magic-modules/commit/dbad345a5d4443e960a6796889d1087801eced5a + I added it as a common diff suppress func, because I think this is quite a common case and couldn't find an existing function for it yet.

I tested this by first reproducing the permadiff by apply-ing via the latest provider version. Then I applied again with my own build, which did not cause the diff.

ramonvermeulen commented on mmv1/third_party/terraform/tpgresource/common_diff_suppress.go:
```diff
@@ -22,6 +22,11 @@ func EmptyOrFalseSuppressBoolean(k, old, new string, d *schema.ResourceData) boo
 
+func EmptyStringOrNilDiffSuppress(k, old, new string, d *schema.ResourceData) bool {
+	o, n := d.GetChange(k)
+	return (o == nil || o == "") && (n != nil && n != "")
```
What is yaqs/3061527455379488768? I manually tested this and it resolves the permadiff for me. Why would you like to extend the validation to check if `old == "0"`? In that scenario there is no diff suppress required right? Because if old == 0 the `preferred_value` was already correct in state and no permadiff would occur?

ScottSuarez replied:
Based on the name `EmptyStringOrNilDiffSuppress`, it seems like it should suppress diffs only when going from empty/nil to empty/nil (treating them as equivalent). But instead, it's suppressing any change from empty/nil to ANY non-empty value.

This is very broad diff suppress and essentially ignoring API value value when empty. That's not what diff suppress is typically used for as we want to target the specific scenario in question.

| would prefer we do something specific to the resource. Like

```
func QuotaPreferredValueDiffSuppress(k, old, new string, d *schema.ResourceData) bool {
	o, n := d.GetChange(k)
	
	// API doesn't return preferred_value when it's "0", so suppress the diff when:
	// - Old (from API) is empty/nil and New (from config) is "0"
	// - OR Old (from API) is "0" and New (from config) is empty/nil (for safety)
	oldEmpty := o == nil || o == ""
	newEmpty := n == nil || n == ""
	
	return (oldEmpty && n == "0") || (o == "0" && newEmpty)
}
```

This function can be specific in the resource file directly by adding it to the constants import like so

https://github.com/ScottSuarez/magic-modules/blob/c5c6f66b096d3e24c16f84e3ffd15acfd3f37fa6/mmv1/products/accesscontextmanager/ServicePerimeter.yaml#L66
